src/classes/RD2_OpportunityService.cls

```diff
@@ -0,0 +1,210 @@
+        Set<Id> createdOppIds = new Set<Id>();
+        Integer oppItem = 0;
+        //Create a list of ContactIds and contactmap used in these RecurringDonations
+        List<Id> contactIds = new List<Id>();
+        Map<Id,Contact> contactMap;
+        Map<Id,String> rdPaymentMethod = new Map<Id,String>();
+        List<Database.SaveResult> badSRs = new List<Database.SaveResult>();
+        List<Opportunity> badOpps = new List<Opportunity>();
+
+        for (npe03__Recurring_Donation__c r : rdsWithCloseDates.keyset()) {
+            if (r.npe03__Contact__c != null) {
+                contactIds.add(r.npe03__Contact__c);
+            }
+            rdPaymentMethod.put(r.Id,r.PaymentMethod__c);
+
+        }
+        contactMap = getContacts(contactIds);
+
+        oppsToInsert = getInstallmentOpportunities(rdsWithCloseDates, contactMap);
+
+        if (oppsToInsert.size() > 0) {
+            for (Database.SaveResult s : insertOpportunities(oppsToInsert)) {
+                if (s.isSuccess() == true) {
+                    //get all the new OppIds into a list
+                    createdOppIds.add(s.getId());
+
+                } else {//handle errors on insert
+                    badSRs.add(s);
+                    badOpps.add(oppsToInsert[oppItem]);
+                }
+                oppItem++;
+            }
+            //Confirm if the error handling is needed due to bulk operations
+            if (!badSRs.isEmpty()) {
+               ERR_Handler.Errors errors = ERR_Handler.getErrors(badSRs, badOpps);
+               ERR_Handler.processErrors(errors.errorRecords, ERR_Handler_API.Context.RD.name());
+            }
+        }
+        updatePaymentRecordOnCreatedOpp(createdOppIds, rdPaymentMethod);
+    }
+
+    /***
+    * @description Gets one or more Opportunity records for a specified set of Recurring Donations with close dates
+    * @param rdsWithCloseDates the Map of Recurring Donations with close dates
+    * @param contactMap the Map of contact with account ids.
+    * @return the list of Opportunities to be created.
+    */
+    private List<Opportunity> getInstallmentOpportunities(Map<npe03__Recurring_Donation__c, List<Date>> rdsWithCloseDates, Map<Id,Contact> contactMap) {
+
+        List<Opportunity> oppsToInsert = new List<Opportunity>();
+        npe03__Recurring_Donations_Settings__c rds = UTIL_CustomSettingsFacade.getRecurringDonationsSettings();
+
+        for (npe03__Recurring_Donation__c r : rdsWithCloseDates.keySet()) {
+
+            for(Integer i = 0;i< rdsWithCloseDates.get(r).size();i++) {
+                Opportunity opp = new Opportunity();
+                opp.Name = r.npe03__Donor_Name__c + ' ' + r.npe03__Amount__c.format() + ' ' + rdsWithCloseDates.get(r)[i].format() + ' ' + System.Label.stgNavRecurringDonations;
+
+                if ( r.npe03__Organization__c != null && String.isBlank(r.npe03__Contact__c)) {
+                    opp.AccountId = r.npe03__Organization__c;
```

Comment:
When the RD.Organization field has an Organization Account, RD's support having any Contact also attached to the RD. Because we already have validation logic in place to ensure that the Organization and Contact fields are correct I don't think there is a need to query the Contact object at all to make these determinations here.

You can simply assign the AccountId and PrimaryContact fields direct from the RD without the additional queries, maps or conditional checks. It would definitely simply the logic here.
```
opp.AccountId = rd.npe03__Organization__c;
opp.Primary_Contact__c rd.npe03__Contact__c;
```